font_collector/ass/abc_ass_document.py

```diff
@@ -0,0 +1,362 @@
+    AssValidTagResetStyle,
+    AssValidTagWrapStyle,
+    parse_line,
+    WrapStyle,
+)
+from copy import deepcopy
+from typing import Dict, List, Optional, Tuple
+
+
+__all__ = ["ABCAssDocument"]
+
+class ABCAssDocument(ABC):
+    """
+    You can extend this class. For an example, see the ass_document.py file.
+    This class is used to parse an .ass file and get all the styles it uses.
+    """
+
+    @abstractmethod
+    def _get_sub_wrap_style(self) -> Optional[WrapStyle]:
+        """
+        Returns:
+            The subtitle WrapStyle.
+            If the subtitle doesn't contain an WrapStyle, it return None
+        """
+        pass
+
+    def get_sub_wrap_style(self) -> WrapStyle:
+        """
+        Returns:
+            The subtitle WrapStyle.
+        """
+
+        sub_wrap_style = self._get_sub_wrap_style()
+        if sub_wrap_style is None:
+            sub_wrap_style = WrapStyle.SMART_TOP
+        return sub_wrap_style
+
+
+    @abstractmethod
+    def get_nbr_style(self) -> int:
+        """
+        Returns:
+            The number of styles.
+        """
+        pass
+
+
+    def __verify_if_style_exist(self, i: int) -> None:
+        if i >= self.get_nbr_style():
+            raise ValueError(f"There isn't any style at the index {i}. There is only {self.get_nbr_style()} style(s)")
+
+
+    @abstractmethod
+    def _get_style(self, i: int) -> Tuple[str, str, bool, bool]:
+        pass
+
+    def get_style(self, i: int) -> Tuple[str, str, bool, bool]:
+        """
+        Args:
+            i: Index of the line.
+        Returns:
+            An Tuple formatted like this: style_name, font_name, is_bold, is_italic
+        """
+        self.__verify_if_style_exist(i)
+        return self._get_style(i)
+
+
+    @abstractmethod
+    def get_nbr_line(self) -> int:
+        """
+        Returns:
+            The number of lines.
+        """
+        pass
+
+
+    def __verify_if_line_exist(self, i: int) -> None:
+        if i >= self.get_nbr_line():
+            raise ValueError(f"There isn't any line at the index {i}. There is only {self.get_nbr_line()} line(s)")
+
+
+    @abstractmethod
+    def _get_line_style_name(self, i: int) -> str:
+        pass
+
+    def get_line_style_name(self, i: int) -> str:
+        """
+        Args:
+            i: Index of the line.
+        Returns:
+            The style name of the line.
+        """
+        self.__verify_if_line_exist(i)
+        line_style_name = self._get_line_style_name(i)
+        # - * https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/STS.cpp#l1490
+        # - tabulation and space : https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/STS.cpp#l1479
+        #  VSFilter set style name to default if empty: https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/STS.cpp#l1892
+        line_style_name = line_style_name.lstrip("\t ").lstrip("*")
+        if len(line_style_name) == 0 or line_style_name.lower() == "default":
```

Comment:
Add link to https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/STS.cpp#l1491 to explain why we don't care about the case of line_style_name if it is equals to "default"